### ISSUE TITLE
Implement track-with-two-round-lap (pkg:laptrack) — verification done, but not the first package-tier skill

### ISSUE THREAD

**Issue by jiyuuchc** [jiyuuchc/biopb#669]
From the skills roadmap (`biopb-site/notes/skills-roadmap.md`, Tier A):

> | `track-with-two-round-lap` | Jaqaman u-track (2008), TrackMate | Round 1 frame-to-frame LAP; round 2 **track-segment** linking for gap closing, merges, splits — the block-structured cost matrix most reimplementations get wrong | `pkg:laptrack` (alt `btrack`) + inline cost/table glue |

This issue records the verification done for that entry, so whoever writes the
body starts from measured facts rather than the roadmap's one-liner.

**Scheduling note: not the first one to implement.** This skill takes a
*coordinate dataframe*, not pixels — `laptrack.predict_dataframe(df,
coordinate_cols, frame_col)`. Detection happens upstream. As the first exercise
of the package tier that is a weak demonstration: it never touches the tensor
read path, the viewer, or dask, which is most of what a biopb skill has to get
right. Better to prove the tier on a skill with image in and image out, then come
back to this one. `deconvolve-with-theoretical-psf` (`pkg:psfmodels`, +1 package,
clean) is the obvious swap — see the end of this issue.

## Why `laptrack` and not something else

I screened 23 candidate packages against the #667 satisfiability gate — the exact
criterion from `test_satisfiability.py`: a package passes iff installing it moves
nothing already in the environment backwards.

**All 23 passed.** That is worth stating on its own: the gate is a floor that
catches a specific pathology, not a high bar. `basicpy` and `m2stitch` were
unusual, not typical. Both were re-checked and both still fail (basicpy moves
numpy 2.3.5, pandas 3.0.3 *and* scipy 1.18.0 backwards; m2stitch takes pandas
3.0.3 → 2.3.3), so the gate's founding examples are still live.

Against that field `laptrack` is the cleanest available first case:

| | |
|---|---|
| install plan | `+ laptrack==0.17.1` — **one package, nothing else** |
| wheel | `py3-none-any` — universal, so all five CI cells give the same answer |
| requires-python | `>=3.10,<3.15` — covers 3.10/3.11/3.12 with headroom |
| license | BSD-3-Clause |
| dependencies | **all already installed**: networkx 3.6.1, pooch 1.9.0, click 8.4.1, pydantic 2.13.4, scikit-learn 1.9.0, scikit-image 0.26.0, pandas 3.0.3, scipy 1.18.0, numpy 2.3.5 |

That last row is why it is a good *first* case and not merely an acceptable one:
a green result means something, because no fat resolution is masking a near-miss.
Note `scipy!=1.15.*` among its pins — we are on 1.18 and clear, but it is a
reminder that the answer is environment-dependent and the gate has to keep asking.

## The round-2 claim checks out

`laptrack/_cost_matrix.py::build_segment_cost_matrix` assembles the
block-structured segment matrix directly — gap-closing (M×M), splitting and
merging blocks, with the `b`/`d`/`b'`/`d'` alternative costs named after the
paper's own variables. The source cites *Jaqaman et al. 2008 Nat Methods* and
cross-references TrackMate's `JaqamanSegmentCostMatrixCreator`. The roadmap's
"the block-structured cost matrix most reimplementations get wrong" is accurate,
and laptrack is not one of the ones getting it wrong.

## API surface a body would quote

Verified by unpacking the 0.17.1 wheel (nothing installed into the shared venv).

- `LapTrack` is a **pydantic model**; parameters are fields, not constructor args
  in the usual sense.
- `predict_dataframe(df, coordinate_cols, frame_col="frame")` returns a
  **3-tuple**: `(track_df, split_df, merge_df)`.
- `track_df` is `set_index(["__frame", "__index"])` — a **MultiIndex**, not a flat
  frame — carrying `track_id` and `tree_id` alongside the coordinate columns.
- `split_df` / `merge_df` carry `parent_track_id` / `child_track_id`.
- `predict()` is the lower-level path and returns a networkx graph.

## Four gotchas — these are the skill's actual content

1. **Cutoffs are squared distances.** The default metric is `sqeuclidean` and
   `gap_closing_cutoff` defaults to `15**2`. Anyone writing `gap_closing_cutoff=15`
   means 15 px and silently gets 15 px² ≈ 3.9 px — four times too strict, no error.
   This is exactly the class of silent failure the package tier exists to prevent.
2. **Splitting and merging are OFF by default.** `splitting_cutoff` and
   `merging_cutoff` both default to `False`. Out of the box you get round 1 plus
   gap closing only. A skill called "two-round LAP" that does not explicitly enable
   them ships a title that misdescribes what the body does.
3. **`track_id` is not `tree_id`.** `track_id` breaks at a division; `tree_id`
   keeps the lineage. Choosing wrong downstream either loses lineage or
   double-counts a dividing cell — and both look plausible in a results table.
4. **`laptrack.__version__` is `"0.17.0"` while the distribution is `0.17.1`.**
   This one is a *runtime* bug, not an authoring one — see below.

## Two findings that outlive this skill

**`__version__` is not a reliable resolver for a `pkg:` token.**
`biopb-mcp/docs/skill-interface.md` §3g specifies that a `pkg:<name>[>=v]` token
resolves to "a real ImportError or a real `__version__`". laptrack ships
`__version__ = "0.17.0"` in a 0.17.1 distribution, so an agent checking
`pkg:laptrack>=0.17.1` reads 0.17.0 and reports the requirement **unmet on a
correctly installed package**. `importlib.metadata.version()` reads the
distribution metadata and gets it right. This is a small change to §3g and to
`guide://kernel`, and it is the first evidence that the rule as written is wrong —
which is an argument for landing the package tier at all, since only a real
package could have surfaced it.

**The satisfiability gate has a blind spot the contract layer should cover.**
`stardist` passes cleanly (`+ csbdeep==0.8.2, + stardist==0.9.2`, nothing moved)
and installs **no tensorflow** — csbdeep declares TF only under a `[tf1]` extra.
So it satisfies the gate, damages nothing, and `import stardist` fails at runtime.
The gate answers *"can this be installed without damage"*, never *"will this
work"*. The second question is what the contract layer (§3, `test_contracts.py`)
was for — written entirely for `flatfield-and-stitch-tiles` and deleted with it in
#667. Whichever package-tier skill lands first should bring it back, now with two
things to assert: that the declared package **imports**, and that the signatures
the body quotes still look that way.

## Suggested first package-tier skill instead

`deconvolve-with-theoretical-psf` (roadmap Tier A, `pkg:psfmodels`) — image in,
image out, so it exercises the tensor read path, memory budgeting, dask and the
viewer, none of which this skill touches. Screened just now: **`+ psfmodels`, one
package, nothing moved** — the same clean profile as laptrack. Gibson–Lanni PSF
from NA/λ/RI/spacing is also squarely the "absent from an LLM's training data"
material the tier is meant for, and the failure it prevents (a Gaussian stand-in
for a real PSF, isotropic sampling on anisotropic voxels) is silent.

Other image-input candidates screened clean, for the record: `dask-image` (+3),
`skan` (+3), `scikit-fmm` (+1), `pyclesperanto` (+2), `porespy` (+4),
`aicsshparam` (+6). `bigfish` does not resolve at all on this Python.

**Comment by jiyuuchc** [comment]
## Correction: withdraw the `psfmodels` suggestion

The "suggested first package-tier skill instead" section above recommends
`deconvolve-with-theoretical-psf` on the strength of a clean satisfiability
screen. **That screen was not sufficient, and psfmodels should not be the first
one.**

`psfmodels` 0.3.3 (latest) publishes wheels only up to **cp311**. On Python 3.12
— three of the five CI cells, and the version this venv runs — there is no usable
wheel, so installing it compiles a C++ extension from the sdist:

```
$ uv pip install --dry-run --no-build psfmodels
  × Because all versions of psfmodels have no usable wheels ... unsatisfiable
```

`--dry-run` alone accepts an sdist without attempting the build, which is why it
read as a pass. That gap in the gate is now #670. It is also **GPL-3.0** against
biopb's MIT — no viral effect, since biopb neither links nor redistributes it and
the user installs it themselves, but that should be a deliberate choice rather
than an accident.

The API is genuinely good material, so recording it here rather than losing it —
lengths are **all in microns** (`wvl=0.6` is 600 nm, and an agent writing
`wvl=488` gets a silently catastrophic PSF); the default `model` is
**`"vectorial"` (Aguet 2009), not Gibson–Lanni** — Gibson–Lanni is
`model="scalar"`, so the roadmap entry naming Gibson–Lanni needs one of the two
corrected; `normalize=True` normalises the **max** to 1, not the sum, which is
wrong for Richardson–Lucy unless renormalised; and `z` as an int versus a sequence
silently changes whether `dz` is used at all.

## Revised candidates, filtered on wheels and licence

| package | wheel on 3.12 | licence | skill |
|---|---|---|---|
| **`dask-image`** | yes | BSD | `segment-large-image-tiled` |
| `skan` | yes | BSD | `skeleton-network-analysis` |
| `pyclesperanto` | yes | BSD | needs an OpenCL runtime — risky |
| `pystackreg` | yes | "Free To Use But Restricted" | `drift-correction` |
| `aicsshparam` | yes | non-commercial only | disqualifying for a shipped skill |
| `scikit-fmm` | **no** | BSD | same sdist problem as psfmodels |
| `psfmodels` | **no** | GPL-3.0 | withdrawn, above |

**`dask-image` is the revised pick.** Image in, labels out; `ndmeasure.label`
links connected components across dask block boundaries, which is the actual
algorithm the skill teaches and the thing a naive `map_overlap` gets wrong; BSD;
real wheels on every cell; and it exercises the dask and tensor read paths harder
than anything else on the list — which was the reason for deferring this issue's
own skill in the first place.

Its API has not been verified to the depth laptrack's was in this issue. That
should happen before it is committed to.
